### DISCUSSION TITLE
Transaction builder with Extended keys (HD Wallet)

### DISCUSSION THREAD

**Discussion by CardanoDur** [Python-Cardano/pycardano#99] [Q&A] [answered]
Got the transaction builder to work great with payment keys generated by cardano-cli so for my next test I wanted to use extended keys.  Able to bring in the signing key just fine, it's the extended validation key that I can't figure out a way to get it witnesses to add to the builder.  

The example [Issue 2](https://github.com/cffls/pycardano/issues/2) shows when building out a raw transaction, but didn't find an example using the builder.  I wouldn't have the transaction body to hash it, to build the payment signature and eventually the witnesses.

Is this possible and does an example exist?

Thanks,
Jason

**Comment by cffls** [comment]
Do you already have an extended signing key? If so, you can simply pass it to the method `builder_and_sign` of txbuilder like you did for regular keys, as documented [here](https://github.com/cffls/pycardano/blob/main/pycardano/txbuilder.py#L1139).

**Reply by CardanoDur** [reply]
Yeah, that's what I did initially, but got a MissingVKeyWitnessesUTXOW error.  I assumed it was due to missing witnesses based on the validation key

**Reply by CardanoDur** [reply]
When I look at the signed TX from the build_and_sign method, the vkey cborHex is not extended.  It's just a partial of the extended key

**Reply by cffls** [reply]
That means the extended signing key you were using didn't match the utxo you were trying to spend. You can verify by checking the hash of your public key with `your_extended_signing_key.to_verification_key().hash()`, and compare its output with the payment part of the address of the utxo you are trying to spend (`utxo_to_spend.output.address.payment_part`).

**Reply by CardanoDur** [reply]
I'm sure it's not a surprise that they don't match.  Today is one of those days.  I know my mistake, I grabbed an address from daedalus after putting the seed phrase into it, obviously that address doesn't match the utxos.  Unfortunately now, after figuring out my mistake, I can't check b/c I'm getting a 404 on the builder.build_and_sign.  Blockfrost isn't showing an issue either

**Reply by CardanoDur** [reply]
Well, that wasn't it either.  I know my public key off the root is correct, it matches what Daedalus has as my wallet public key.  But for some reason I can't get the signing key right.  Can't get the hash to match the UTXO.  My mind is mush...

**Comment by CardanoDur** [comment, marked as answer]
@cffls Thank you for your help, my issue was going for a shortcut.  I didn't generate an address based off of the payment and staking keys, I just imported to Daedalus and grabbed an address then funded.  Once I generated an address then funded that address (same stake, so it was a wasted internal transfer for the most part), the code worked great

**Reply by cffls** [reply]
Nice, thanks for the update!

**Reply by cffls** [reply]
Just found that you started sponsoring me. Really appreciate it. Thank you so much! ❤️
